aten/src/ATen/functorch/BatchedFallback.cpp

```diff
@@ -396,6 +396,120 @@ void batchedTensorForLoopFallback(const c10::OperatorHandle& op, torch::jit::Sta
+    const auto& ivalue = arguments[idx];
+    if (!ivalue.isTensor()) {
+      continue;
+    }
+    const auto& tensor = ivalue.toTensor();
+    if (!tensor.defined()) {
+      continue;
+    }
+    const auto* batched = maybeGetBatchedImpl(tensor);
+    if (!batched) {
+      continue;
+    }
+    batched_tensor_inputs.push_back(tensor);
+    batched_tensor_inputs_position.push_back(idx);
+  }
+  TORCH_INTERNAL_ASSERT(!batched_tensor_inputs.empty());
+
+  std::vector<std::vector<Tensor>> unbound;
+  for (auto iter = batched_tensor_inputs.begin(); iter != batched_tensor_inputs.end(); ++iter) {
+    auto *batched_impl = maybeGetBatchedImpl(*iter);
+    TORCH_INTERNAL_ASSERT(batched_impl->value().is_nested() || batched_impl->bdim() == 0,
+        "Fallback not supported for mixed nested / non-nested arguments without bdim=0");
+    c10::impl::ExcludeDispatchKeyGuard guard(DispatchKey::BatchedNestedTensor);
+    auto this_unbound = batched_impl->value().unbind();
+    if (unbound.size() > 0) {
+      TORCH_INTERNAL_ASSERT(unbound.front().size() == this_unbound.size(),
+          "Fallback not supported for differently-sized nested arguments");
+    }
+    unbound.push_back(this_unbound);
+  }
+
+  const auto num_components = unbound.front().size();
+  std::vector<Tensor> output_shards(num_components * num_returns);
+  for (const auto component_idx : c10::irange(0, num_components)) {
+    auto batched_idx = 0;
+    auto batched_tensor_inputs_pos_iter = batched_tensor_inputs_position.begin();
+    for (const auto arg_idx : c10::irange(0, num_arguments)) {
+      // We assume that torch::jit::Stack is backed by vector<IValue> for
+      // simplicity. When that is not the case, this code should be updated.
+      const auto& argument = (*stack)[arguments_begin + arg_idx];
+      if (batched_tensor_inputs_pos_iter == batched_tensor_inputs_position.end()
+          || (int64_t)arg_idx != *batched_tensor_inputs_pos_iter) {
+        // argument isn't a BatchedTensor
+        torch::jit::push(stack, argument);
+        continue;
+      }
+      // argument is a BatchedTensor
+      torch::jit::push(stack, unbound[batched_idx][component_idx]);
+      ++batched_idx;
+      ++batched_tensor_inputs_pos_iter;
+    }
+
+    c10::impl::ExcludeDispatchKeyGuard guard(DispatchKey::BatchedNestedTensor);
+    op.callBoxed(stack);
+
+    // Store the result into `output_shards`. See NOTE: [Output shards layout]
+    // to learn about the details of how we store the shards.
+    const auto returns = torch::jit::last(stack, num_returns);
+    for (const auto return_idx : c10::irange(0, returns.size())) {
+      output_shards[num_components * return_idx + component_idx] = returns[return_idx].toTensor();
+    }
+    torch::jit::drop(stack, num_returns);
+  }
+
+  // For each output Tensor, stack the shards of the tensor together to form a nested return
+  // TODO: Determine when the output needs to be nested and when it can be non-nested?
```

Comment:
I guess with something like torch.sum() you know the output shouldn't be nested. How is this case handled for regular NestedTensors? e.g. does torch.sum(NestedTensor) return a regular Tensor?